Solved for Issue #64321: handle user passing unroll-factor <=0

SwapnilGhanshyala opened the pull request.

bondhugula commented:
Please use an appropriate commit title: instead of "Solved ...", use "[MLIR][Affine] Fix crash with an invalid unroll factor", etc. 
The commit is still missing a commit summary.

SwapnilGhanshyala commented:
Updated the commit message and the commit summary.

bondhugula commented on mlir/lib/Dialect/Affine/Utils/LoopUtils.cpp:
```diff
@@ -23,6 +23,7 @@
 #include "mlir/Dialect/SCF/IR/SCF.h"
 #include "mlir/IR/IRMapping.h"
 #include "mlir/IR/IntegerSet.h"
+#include "mlir/Support/LogicalResult.h"
```
You don't need this include I think - it's transitively covered.

bondhugula commented on mlir/lib/Dialect/Affine/Utils/LoopUtils.cpp:
```diff
@@ -1016,8 +1017,12 @@ LogicalResult mlir::affine::loopUnrollByFactor(
     function_ref<void(unsigned, Operation *, OpBuilder)> annotateFn,
     bool cleanUpUnroll) {
-  assert(unrollFactor > 0 && "unroll factor should be positive");
-
+  if (unrollFactor <= 0) {
+    return emitError(
+        forOp.getLoc(),
+        "'-affine-loop-unroll=unroll-factor' cannot be less than or "
+        "equal to zero; pass aborted!");
+  }
```
The assert is fine. The utility shouldn't typically emit a message in this situation. The 0 unroll factor should be checked by the user of this utility, which would be the pass, given the crash issue you are addressing. Please put the check and emit message in the pass.

bondhugula reviewed (changes requested):
Commit titles shouldn't have full stops - please drop it.

SwapnilGhanshyala commented:
Changes Addressed:
1) check user error in pass itself
2) Updated commit title and message:
 [MLIR][Affine] Fix crash with invalid unroll factor (Issue #64321)

Updated LoopUnroll pass to check if unrollFactor passed by the user is invalid, i.e., less than or equal to
zero. If so, emit appropriate message and signal pass failure.

Reporting Issue is #64321

bondhugula commented on mlir/lib/Dialect/Affine/Transforms/LoopUnroll.cpp:
```diff
@@ -94,7 +94,12 @@ void LoopUnroll::runOnOperation() {
   if (func.isExternal())
     return;
-
+  if (unrollFactor <= 0) {
```
This is an `unsigned`. `== 0` would be the right check.

bondhugula replied:
Fix message accordingly as well.
